README.md

# IEC 62056-21

[![CircleCI](https://circleci.com/gh/pwitab/iec62056-21/tree/master.svg?style=svg)](https://circleci.com/gh/pwitab/iec62056-21/tree/master)
[![Coverage Status](https://coveralls.io/repos/github/pwitab/iec62056-21/badge.svg)](https://coveralls.io/github/pwitab/iec62056-21)

A Python library for IEC 62056-21, Direct Local Data Exchange of Energy Meters. 
Former IEC 61107 or IEC 1107

## Installation

We only support python 3.6+

Install via pip:

```
pip install iec62056-21
```

## About IEC 62056-21

IEC 62056-21 (earlier IEC 61107 or sometimes just IEC 1107, is an international 
standard for a computer protocol to read utility meters. It is designed to operate 
over any media, including the Internet. A meter sends ASCII (in modes A..D) or 
HDLC (mode E) data to a nearby hand-held unit (HHU) using a serial port. The physical 
media are usually either modulated light, sent with an LED and received with a 
photodiode, or a pair of wires, usually modulated by a 20mA current loop. The protocol 
is usually half-duplex.


## Limitations of this library.

* At the moment we only support Mode C.
* We assume that only protocol mode Normal is used.

## Example usage:

Reading a meter using a optical usb probe via the D0-interface.

```python
from iec62056_21.client import Iec6205621Client

client = Iec6205621Client.with_serial_transport(port='/dev/tty_something')
client.connect()
print(client.standard_readout())

```


Reading a meter over an internet connection.
```python
from iec62056_21.client import Iec6205621Client

client = Iec6205621Client.with_tcp_transport(address=('192.168.0.1', 8000), device_address='12345678', password='00000000')
client.connect()
print(client.standard_readout())
```


## Derivative protocols

Some manufacturer are using a derivative protocol to IEC 62056-21. They comply with 
most things but might for example not use the access request features according to 
standard or they have a slightly different flow in command execution

This library can be used with some of them. You just need to be aware of the differences.
We provide special handlers for some unique parts that is included in this library. 
They might be split into separate libraries in the future.

### LIS-200

A protocol for Elster devices. Main difference is that they have the concept of 
locks instead of password and instead of answering the password request you need to 
write the password to a certain register.

## Development

This library is developed by Palmlund Wahlgren Innovative Technology AB in Sweden and 
is used in our multi utility AMR solution: [Utilitarian](https://docs.utilitarian.io)

## Contributing

*   Check for open issues or open a fresh issue to start a discussion around a feature 
    idea or a bug.
*   Fork the repository on GitHub to start making your changes to the master branch (or 
    branch off of it).
*   Write a test which shows that the bug was fixed or that the feature works as expected.
*   Send a pull request and bug the maintainer until it gets merged and published.


HISTORY.md

# Changelog
All notable changes to this project will be documented in this file.


The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Calendar Versioning](https://calver.org/)


## Unreleased

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## [21.1.0]  (2021-04-26)

### Fixed

* Fixed problem accessing programing mode. `#7`
* Fixed problem setting different start baud rate for use with for example RS485. `#9`
* A simpler code example in the README. `#6`

## v.0.0.2 (2019-06-12)

### Fixed

Issue `#2` Dependency problems when installing.

## v0.0.1 (2019-06-04)

Initial implementation of IEC 62056-21 with focus on supporting LIS 200 derivative 
protocol.

tests/test_bcc.py
import pytest

from iec62056_21.utils import calculate_bcc, add_bcc


class TestBcc:
    def test_bcc_bytes1(self):
        data = bytes.fromhex("01573202433030332839313033323430393232333929031b")
        correct_bcc = chr(data[-1]).encode("latin-1")
        bcc = calculate_bcc(data[1:-1])
        assert bcc == correct_bcc

    def test_bcc_bytes_2(self):
        data = b"\x01P0\x02(1234567)\x03P"
        correct_bcc = chr(data[-1]).encode("latin-1")
        bcc = calculate_bcc(data[1:-1])
        assert bcc == correct_bcc

    def test_bcc_string(self):
        data = "\x01P0\x02(1234567)\x03P"
        correct_bcc = data[-1]
        bcc = calculate_bcc(data[1:-1])
        assert bcc == correct_bcc

    def test_add_bcc1(self):
        data = "\x01P0\x02(1234567)\x03"
        correct_data = "\x01P0\x02(1234567)\x03P"
        with_bcc = add_bcc(data)
        assert with_bcc == correct_data


tests/test_data.py
import pytest

from iec62056_21 import messages, exceptions, constants


class TestDataSets:

    data_set_with_unit = "3.1.0(100*kWh)"
    data_set_without_unit = "3.1.0(100)"
    not_valid_data = '"Tralalalala'

    def test_from_string_with_unit(self):
        ds = messages.DataSet.from_representation(self.data_set_with_unit)
        assert ds.value == "100"
        assert ds.address == "3.1.0"
        assert ds.unit == "kWh"

    def test_from_bytes_with_unit(self):
        ds_bytes = self.data_set_with_unit.encode("latin-1")
        ds = messages.DataSet.from_representation(self.data_set_with_unit)
        assert ds.value == "100"
        assert ds.address == "3.1.0"
        assert ds.unit == "kWh"

    def test_to_string_with_unit(self):
        ds = messages.DataSet(value="100", address="3.1.0", unit="kWh")
        assert ds.to_representation() == self.data_set_with_unit

    def test_to_byte_with_unit(self):
        ds = messages.DataSet(value="100", address="3.1.0", unit="kWh")
        assert ds.to_bytes() == self.data_set_with_unit.encode(constants.ENCODING)

    def test_from_string_without_unit(self):
        ds = messages.DataSet.from_representation(self.data_set_without_unit)
        assert ds.value == "100"
        assert ds.address == "3.1.0"
        assert ds.unit is None

    def test_from_bytes_without_unit(self):
        ds_bytes = self.data_set_without_unit.encode("latin-1")
        ds = messages.DataSet.from_representation(self.data_set_without_unit)
        assert ds.value == "100"
        assert ds.address == "3.1.0"
        assert ds.unit is None

    def test_to_string_without_unit(self):
        ds = messages.DataSet(value="100", address="3.1.0", unit=None)
        assert ds.to_representation() == self.data_set_without_unit

    def test_to_byte_without_unit(self):
        ds = messages.DataSet(value="100", address="3.1.0", unit=None)
        assert ds.to_bytes() == self.data_set_without_unit.encode(constants.ENCODING)

    def test_invalid_data(self):
        with pytest.raises(exceptions.Iec6205621ParseError):
            ds = messages.DataSet.from_representation(self.not_valid_data)


class TestBase:
    def test_to_bytes(self):
        with pytest.raises(NotImplementedError):
            messages.Iec6205621Data().to_bytes()

    def test_from_bytes(self):
        with pytest.raises(NotImplementedError):
            messages.Iec6205621Data().from_bytes(b"1235")


class TestDataLine:
    def test_from_representation(self):
        string_data = "12(12*kWh)13(13*kWh)14(14*kwh)"

        dl = messages.DataLine.from_representation(string_data)

        assert len(dl.data_sets) == 3
        assert dl.data_sets[0].value == "12"
        assert dl.data_sets[0].address == "12"
        assert dl.data_sets[0].unit == "kWh"

    def test_to_representation(self):
        dl = messages.DataLine(
            data_sets=[
                messages.DataSet(address="3:14", value="314", unit="kWh"),
                messages.DataSet(address="4:15", value="415", unit="kWh"),
            ]
        )

        rep = dl.to_representation()

        assert rep == "3:14(314*kWh)4:15(415*kWh)"


class TestDataBlock:
    def test_from_representation_single_line(self):
        string_data = "12(12*kWh)13(13*kWh)14(14*kwh)\r\n"

        db = messages.DataBlock.from_representation(string_data)

        assert len(db.data_lines) == 1

    def test_from_representation_several_lines(self):
        string_data = (
            "12(12*kWh)13(13*kWh)14(14*kwh)\r\n"
            "12(12*kWh)13(13*kWh)14(14*kwh)\r\n"
            "12(12*kWh)13(13*kWh)14(14*kwh)\r\n"
        )
        db = messages.DataBlock.from_representation(string_data)

        assert len(db.data_lines) == 3

    def test_to_representation_several_lines(self):

        db = messages.DataBlock(
            data_lines=[
                messages.DataLine(
                    data_sets=[
                        messages.DataSet(address="3:14", value="314", unit="kWh"),
                        messages.DataSet(address="4:15", value="415", unit="kWh"),
                    ]
                ),
                messages.DataLine(
                    data_sets=[
                        messages.DataSet(address="3:14", value="314", unit="kWh"),
                        messages.DataSet(address="4:15", value="415", unit="kWh"),
                    ]
                ),
                messages.DataLine(
                    data_sets=[
                        messages.DataSet(address="3:14", value="314", unit="kWh"),
                        messages.DataSet(address="4:15", value="415", unit="kWh"),
                    ]
                ),
            ]
        )

        assert db.to_representation() == (
            "3:14(314*kWh)4:15(415*kWh)\r\n"
            "3:14(314*kWh)4:15(415*kWh)\r\n"
            "3:14(314*kWh)4:15(415*kWh)\r\n"
        )

        def test_to_representation_single_line(self):
            db = messages.DataBlock(
                data_lines=[
                    messages.DataLine(
                        data_sets=[
                            messages.DataSet(address="3:14", value="314", unit="kWh"),
                            messages.DataSet(address="4:15", value="415", unit="kWh"),
                        ]
                    )
                ]
            )

            assert db.to_representation() == "3:14(314*kWh)4:15(415*kWh)\r\n"


class TestAnswerDataMessage:
    def test_from_representation(self):
        data = "\x023:171.0(0)\x03\x12"

        am = messages.AnswerDataMessage.from_representation(data)

        assert am.data_block.data_lines[0].data_sets[0].value == "0"
        assert am.data_block.data_lines[0].data_sets[0].address == "3:171.0"
        assert am.data[0].value == "0"
        assert am.data[0].address == "3:171.0"

    def test_from_representation_invalid_bcc_raises_value_error(self):
        data = "\x023:171.0(0)\x03\x11"
        with pytest.raises(ValueError):
            am = messages.AnswerDataMessage.from_representation(data)

    def test_to_representation(self):
        db = messages.DataBlock(
            data_lines=[
                messages.DataLine(
                    data_sets=[
                        messages.DataSet(address="3:14", value="314", unit="kWh"),
                        messages.DataSet(address="4:15", value="415", unit="kWh"),
                    ]
                )
            ]
        )

        am = messages.AnswerDataMessage(data_block=db)

        assert am.to_representation() == "\x023:14(314*kWh)4:15(415*kWh)\r\n\x03\x04"


class TestReadoutDataMessage:
    def test_to_representation(self):
        rdm = messages.ReadoutDataMessage(
            data_block=messages.DataBlock(
                data_lines=[
                    messages.DataLine(
                        data_sets=[
                            messages.DataSet(address="3:14", value="314", unit="kWh"),
                            messages.DataSet(address="4:15", value="415", unit="kWh"),
                        ]
                    )
                ]
            )
        )

        assert rdm.to_representation() == '\x023:14(314*kWh)4:15(415*kWh)\r\n!\r\n\x03"'

    def test_from_representation(self):

        rdm = messages.ReadoutDataMessage.from_representation(
            '\x023:14(314*kWh)4:15(415*kWh)\r\n!\r\n\x03"'
        )
        print(rdm)
        assert len(rdm.data_block.data_lines) == 1
        assert len(rdm.data_block.data_lines[0].data_sets) == 2

    def test_invalid_bcc_raises_error(self):
        with pytest.raises(ValueError):
            rdm = messages.ReadoutDataMessage.from_representation(
                "\x023:14(314*kWh)4:15(415*kWh)\r\n!\r\n\x03x"
            )


class TestCommandMessage:
    def test_command_message_to_representation(self):
        cm = messages.CommandMessage(
            command="R",
            command_type="1",
            data_set=messages.DataSet(address="1.8.0", value="1"),
        )

        assert cm.to_representation() == "\x01R1\x021.8.0(1)\x03k"

    def test_from_representation(self):
        data = "\x01P0\x02(1234567)\x03P"
        cm = messages.CommandMessage.from_representation(data)

        assert cm.command == "P"
        assert cm.command_type == "0"
        assert cm.data_set.value == "1234567"
        assert cm.data_set.address is None
        assert cm.data_set.unit is None

    def test_invalid_command_raises_value_error(self):
        with pytest.raises(ValueError):
            cm = messages.CommandMessage(
                command="X",
                command_type="1",
                data_set=messages.DataSet(address="1.8.0", value="1"),
            )

    def test_invalid_command_type_raises_value_error(self):
        with pytest.raises(ValueError):
            cm = messages.CommandMessage(
                command="R",
                command_type="12",
                data_set=messages.DataSet(address="1.8.0", value="1"),
            )

    def test_to_representation_without_data_set(self):
        # like the break command

        break_msg = messages.CommandMessage(
            command="B", command_type="0", data_set=None
        )

        assert break_msg.to_representation() == "\x01B0\x03q"

    def test_from_representation_invalid_bcc_raise_value_error(self):
        data = "\x01P0\x02(1234567)\x03X"
        with pytest.raises(ValueError):
            cm = messages.CommandMessage.from_representation(data)

    def test_for_single_read(self):
        cm = messages.CommandMessage.for_single_read(address="1.8.0")

        assert cm.command == "R"
        assert cm.command_type == "1"
        assert cm.data_set.address == "1.8.0"

        cm = messages.CommandMessage.for_single_read(
            address="1.8.0", additional_data="1"
        )

        assert cm.command == "R"
        assert cm.command_type == "1"
        assert cm.data_set.address == "1.8.0"
        assert cm.data_set.value == "1"

    def test_for_single_write(self):
        cm = messages.CommandMessage.for_single_write(address="1.8.0", value="123")

        assert cm.command == "W"
        assert cm.command_type == "1"
        assert cm.data_set.address == "1.8.0"
        assert cm.data_set.value == "123"


class TestRequestMessage:
    def test_to_representation(self):
        rm = messages.RequestMessage(device_address="45678903")

        assert rm.to_representation() == "/?45678903!\r\n"

    def test_to_representation_without_address(self):
        rm = messages.RequestMessage()

        assert rm.to_representation() == "/?!\r\n"

    def test_from_representation(self):

        in_data = "/?45678903!\r\n"

        rm = messages.RequestMessage.from_representation(in_data)

        assert rm.device_address == "45678903"


class TestAckOptionSelectMessage:
    def test_to_representation(self):
        aosm = messages.AckOptionSelectMessage(mode_char="1", baud_char="5")

        assert aosm.to_representation() == "\x06051\r\n"

    def test_from_representation(self):
        aosm = messages.AckOptionSelectMessage.from_representation("\x06051\r\n")

        assert aosm.mode_char == "1"
        assert aosm.baud_char == "5"


class TestIdentificationMessage:
    def test_to_representation(self):

        im = messages.IdentificationMessage(
            identification="2EK280", manufacturer="Els", switchover_baudrate_char="6"
        )

        assert im.to_representation() == "/Els6\\2EK280\r\n"

    def test_from_representation(self):
        im = messages.IdentificationMessage.from_representation("/Els6\\2EK280\r\n")

        assert im.identification == "2EK280"
        assert im.manufacturer == "Els"
        assert im.switchover_baudrate_char == "6"


tests/test_client.py
import pytest
from iec62056_21 import exceptions, client, transports


class TestIec6205621Client:
    def test_with_no_address_when_required_raises_client_error(self):
        with pytest.raises(exceptions.Iec6205621ClientError):
            c = client.Iec6205621Client.with_tcp_transport(("192.168.1.1", 5000))

    def test_can_create_client_with_tcp_transport(self):
        c = client.Iec6205621Client.with_tcp_transport(
            "192.168.1.1", device_address="00000000"
        )

    def test_no_address_when_required_raises_client_error(self):
        trans = transports.TcpTransport(address=("192.168.1.1", 5000))
        with pytest.raises(exceptions.Iec6205621ClientError):
            c = client.Iec6205621Client(transport=trans)

    def test_can_create_client_tcp_transport(self):
        trans = transports.TcpTransport(address=("192.168.1.1", 5000))
        c = client.Iec6205621Client(transport=trans, device_address="00000000")


iec62056_21/utils.py
from iec62056_21 import constants


def bcc_valid(message):
    bcc = message[-1]
    to_calc = message[:-1]
    calc = add_bcc(to_calc)
    if message == calc:
        return True
    else:
        return False


def add_bcc(message):
    """
    Returns the message with BCC added.
    Data to use starts after STX and ends with but includes ETX
    If there is a SOH in the message the calculation should be done from there.
    """

    if isinstance(message, str):
        _message = message.encode(constants.ENCODING)
        return _add_bcc(_message).decode(constants.ENCODING)
    return _add_bcc(message)


def _add_bcc(message: bytes):
    start_bcc_index = 1
    soh_index = message.find(constants.SOH.encode(constants.ENCODING))
    if soh_index == -1:
        # SOH not found
        stx_index = message.find(constants.STX.encode(constants.ENCODING))
        if stx_index == -1:
            raise IndexError("No SOH or STX found i message")
        start_bcc_index = stx_index + 1
    else:
        start_bcc_index = soh_index + 1

    data_for_bcc = message[start_bcc_index:]
    bcc = calculate_bcc(data_for_bcc)
    return message + bcc


def calculate_bcc(data):
    """
    Calculate BCC.
    """
    if isinstance(data, str):
        _bcc = _calculate_bcc(data.encode(constants.ENCODING))
        return _bcc.decode(constants.ENCODING)

    return _calculate_bcc(data)


def _calculate_bcc(bytes_data: bytes):
    bcc = 0
    for b in bytes_data:
        x = b & 0x7F
        bcc ^= x
        bcc &= 0x7F
    return bcc.to_bytes(length=1, byteorder="big")


def ensure_bytes(data):
    if isinstance(data, str):
        return data.encode(constants.ENCODING)
    elif isinstance(data, bytes):
        return data
    else:
        raise ValueError(f"data:{data!r} cant be converted to bytes")


iec62056_21/exceptions.py
class Iec6205621Exception(Exception):
    """General IEC62056-21 Exception"""


class Iec6205621ClientError(Iec6205621Exception):
    """Client error"""


class Iec6205621ParseError(Iec6205621Exception):
    """Error in parsing IEC62056-21 data"""


class ValidationError(Iec6205621Exception):
    """Not valid data error"""


class TooManyValuesReturned(Iec6205621Exception):
    """If a request for a single value returned more than one value"""


class NoDataReturned(Iec6205621Exception):
    """No data was returned"""


class Iec6206521BaseErrorParser:
    """
    Error messages are contained in DataSets and values without unit. Their format is
    manufacturer specific so the library can only define a way to handle them not the
    exact implementation.
    The DummyErrorParser will we used as standard that ignores all errors.
    An ErrorParser should take an answer response and parse each data set in it to see
    if there is any errors. It should raise appropriate exceptions.
    """

    def __init__(self):
        pass

    def check_for_errors(self, answer_response):
        raise NotImplementedError("check_for_errors must be implemented in subclass")


class DummyErrorParser(Iec6206521BaseErrorParser):
    """
    A Dummy parser that fits in as default. Should be overridden if you want to define
    an error parser.
    """

    def check_for_errors(self, answer_response):
        pass


iec62056_21/client.py
import time
import logging

from iec62056_21 import messages, constants, transports, exceptions

logger = logging.getLogger(__name__)


class Iec6205621Client:
    """
    A client class for IEC 62056-21. Only validated with meters using mode C.
    """

    BAUDRATES_MODE_C = {
        "0": 300,
        "1": 600,
        "2": 1200,
        "3": 2400,
        "4": 4800,
        "5": 9600,
        "6": 19200,
    }
    ALLOWED_MODES = [
        "readout",
        "programming",
        "binary",
        "manufacturer6",
        "manufacturer7",
        "manufacturer8",
        "manufacturer9",
    ]
    MODE_CONTROL_CHARACTER = {
        "readout": "0",
        "programming": "1",
        "binary": "2",
        "manufacturer6": "6",
        "manufacturer7": "7",
        "manufacturer8": "8",
        "manufacturer9": "9",
    }
    SHORT_REACTION_TIME = 0.02
    REACTION_TIME = 0.2

    def __init__(
        self,
        transport,
        device_address="",
        password="00000000",
        battery_powered=False,
        error_parser_class=exceptions.DummyErrorParser,
    ):

        self.transport = transport
        self.device_address = device_address
        self.password = password
        self.battery_powered = battery_powered
        self.identification = None
        self._switchover_baudrate_char = None
        self.manufacturer_id = None
        self.use_short_reaction_time = False
        self.error_parser = error_parser_class()
        self._current_baudrate: int = 300

        if self.transport.TRANSPORT_REQUIRES_ADDRESS and not self.device_address:
            raise exceptions.Iec6205621ClientError(
                f"The transported used ({self.transport}) requires a device address "
                f"and none was supplied."
            )

    @property
    def switchover_baudrate(self):
        """
        Shortcut to get the baud rate for the switchover.
        """
        return self.BAUDRATES_MODE_C.get(self._switchover_baudrate_char)

    def read_single_value(self, address, additional_data="1"):
        """
        Reads a value from an address in the device.

        :param address:
        :param additional_data:
        :return:
        """
        # TODO Can't find documentation on why the additional_data of 1 is needed.
        #  LIS-200 Specific?

        # TODO: When not using the additional data on an EMH meter we get an ack back.
        #   a bit later we get the break message. Is the device waiting?

        request = messages.CommandMessage.for_single_read(address, additional_data)
        logger.info(f"Sending read request: {request}")
        self.transport.send(request.to_bytes())

        response = self.read_response()

        if len(response.data) > 1:
            raise exceptions.TooManyValuesReturned(
                f"Read of one value returned {len(response.data)}"
            )
        if len(response.data) == 0:
            raise exceptions.NoDataReturned(f"Read returned no data")

        logger.info(f"Received response: {response}")
        # Just return the data, not in a list since it is just one.
        return response.data[0]

    def write_single_value(self, address, data):
        """
        Writes a value to an address in the device.

        :param address:
        :param data:
        :return:
        """

        request = messages.CommandMessage.for_single_write(address, data)
        logger.info(f"Sending write request: {request}")
        self.transport.send(request.to_bytes())

        ack = self._recv_ack()
        if ack == constants.ACK:
            logger.info(f"Write request accepted")
            return
        elif ack == constants.NACK:
            # TODO: implement retry and raise proper error.
            raise ValueError(f"Received NACK upon sending {request}")
        else:
            raise ValueError(
                f"Received invalid response {ack} to write request {request}"
            )

    def connect(self):
        """
        Connect to the device
        """
        self.transport.connect()

    def disconnect(self):
        """
        Close connection to device
        """
        self.transport.disconnect()

    def startup(self):
        """
        Initial communication to start the session with the device. Sends a
        RequestMessage and receives identification message.
        """

        if self.battery_powered:
            self.send_battery_power_startup_sequence()
        logger.info("Staring init sequence")
        self.send_init_request()

        ident_msg = self.read_identification()

        # Setting the baudrate to the one propsed by the device.
        self._switchover_baudrate_char = ident_msg.switchover_baudrate_char
        self.identification = ident_msg.identification
        self.manufacturer_id = ident_msg.manufacturer

        # If a meter transmits the third letter (last) in lower case, the minimum
        # reaction time for the device is 20 ms instead of 200 ms.
        if self.manufacturer_id[-1].islower():
            self.use_short_reaction_time = True

    def access_programming_mode(self):
        """
        Goes through the steps to set the meter in programming mode.
        Returns the password challenge request to be acted on.
        """

        self.startup()

        self.ack_with_option_select("programming")

        # receive password request
        pw_req = self.read_response()

        return pw_req

    def standard_readout(self):
        """
        Goes through the steps to read the standard readout response from the device.
        """
        self.startup()
        self.ack_with_option_select("readout")
        logger.info(f"Reading standard readout from device.")
        response = self.read_response()
        return response

    def send_password(self, password=None):
        """
        On receiving the password challenge request one must handle the password
        challenge according to device specification and then send the password.
        :param password:
        """
        _pw = password or self.password
        data_set = messages.DataSet(value=_pw)
        cmd = messages.CommandMessage(command="P", command_type="1", data_set=data_set)
        logger.info("Sending password to meter")
        self.transport.send(cmd.to_bytes())

    def send_break(self):
        """
        Sending the break message to indicate that one wants to stop the
        communication.
        """
        logger.info("Sending BREAK message to end communication")
        break_msg = messages.CommandMessage(
            command="B", command_type="0", data_set=None
        )
        self.transport.send(break_msg.to_bytes())

    def ack_with_option_select(self, mode):
        """
        After receiving the identification one needs to respond with an ACK including
        the different options for the session. The main usage is to control the
        mode. readout, programming, or manufacturer specific. The baudrate change used
        will be the one proposed by the device in the identification message.

        :param mode:
        """
        # TODO: allow the client to suggest a new baudrate to the devices instead of
        #  the devices proposed one.

        mode_char = self.MODE_CONTROL_CHARACTER[mode]

        ack_message = messages.AckOptionSelectMessage(
            mode_char=mode_char, baud_char=self._switchover_baudrate_char
        )
        logger.info(f"Sending AckOptionsSelect message: {ack_message}")
        self.transport.send(ack_message.to_bytes())
        self.rest()
        self.transport.switch_baudrate(
            baud=self.BAUDRATES_MODE_C[self._switchover_baudrate_char]
        )

    def send_init_request(self):
        """
        The init request tells the device they you want to start a session with it.
        When using the optical interface on the device there is no need to send the
        device address in the init request since there can be only one meter.
        Over TCP or bus-like transports like RS-485 you will need to specify the meter
         you want to talk to by adding the address in the request.

        """
        request = messages.RequestMessage(device_address=self.device_address)
        logger.info(f"Sending request message: {request}")
        self.transport.send(request.to_bytes())
        self.rest()

    def read_identification(self):
        """
        Properly receive the identification message and parse it.
        """

        data = self.transport.simple_read(start_char="/", end_char="\x0a")

        identification = messages.IdentificationMessage.from_bytes(data)
        logger.info(f"Received identification message: {identification}")
        return identification

    def send_battery_power_startup_sequence(self, fast=False):
        """
        Battery powered devices require a startup sequence of null bytes to
        activate
        There is a normal and a fast start up sequence defined in the protocol.

        Normal:
            Null chars should be sent to the device for 2.1-2.3 seconds with a maximum
            of 0,5 seconds between them.
            After the last charachter the client shall wait 1.5-1,7 seconds until it
            sends the request message

        :param fast:
        """
        if fast:
            raise NotImplemented("Fast startup sequence is not yet implemented")

        timeout = 2.2
        duration = 0
        start_time = time.time()
        logger.info("Sending battery startup sequence")
        while duration < timeout:
            out = b"\x00"
            self.transport.send(out)
            self.rest(0.2)
            duration = time.time() - start_time
        logger.info("Startup Sequence finished")

        self.rest(1.5)

    def _recv_ack(self):
        """
        Simple way of receiving an ack or nack.
        """
        ack = self.transport.recv(1).decode(constants.ENCODING)
        return ack

    def read_response(self, timeout=None):
        """
        Reads the response from a device and parses it to the correct message type.

        :param timeout:
        """
        data = self.transport.read()
        if data.startswith(b"\x01"):
            # We probably received a password challenge
            return messages.CommandMessage.from_bytes(data)
        else:
            response = messages.AnswerDataMessage.from_bytes(data)
            self.error_parser.check_for_errors(response)
            return response

    @property
    def reaction_time(self):
        """
        The device can define two different reaction times. Depending if the third
        letter in the manufacturer ID in the identification request is in lower case the
        shorter reaction time is used.
        """
        if self.use_short_reaction_time:
            return self.SHORT_REACTION_TIME
        else:
            return self.REACTION_TIME

    def rest(self, duration=None):
        """
        The protocol needs some timeouts between reads and writes to enable the device
        to properly parse a message and return the result.
        """

        _duration = duration or (self.reaction_time * 1.25)
        logger.debug(f"Resting for {_duration} seconds")
        time.sleep(_duration)

    @classmethod
    def with_serial_transport(
        cls,
        port,
        device_address="",
        password="00000000",
        battery_powered=False,
        error_parser_class=exceptions.DummyErrorParser,
    ):
        """
        Initiates the client with a serial transport.

        :param port:
        :param device_address:
        :param password:
        :param battery_powered:
        :return:
        """
        transport = transports.SerialTransport(port=port)
        return cls(
            transport, device_address, password, battery_powered, error_parser_class
        )

    @classmethod
    def with_tcp_transport(
        cls,
        address,
        device_address="",
        password="00000000",
        battery_powered=False,
        error_parser_class=exceptions.DummyErrorParser,
    ):
        """
        Initiates the client with a TCP Transport.

        :param address:
        :param device_address:
        :param password:
        :param battery_powered:
        :return:
        """
        transport = transports.TcpTransport(address=address)
        return cls(
            transport, device_address, password, battery_powered, error_parser_class
        )
